### ISSUE TITLE
Failing miserably to get Postman to POST a SOAP envelope to a Method

### ISSUE THREAD

**Issue by VorTechS** [CoreWCF/CoreWCF#49]
Firstly, let me start by saying, I have absolutely no idea what I'm doing really!  I'm new to .NET Core and I've never touched MVC.  I AM familiar with high-level WCF, creating services and bindings but nothing anywhere near as low-level as this project.  So I am delving into things quickly trying to learn as I go along.

So my aim, for this test project, is to build an application that leverages Controllers, but also SOAP.  (Yes it's not normal I know, but I just want to give myself an architectural foundation whereby rather than having relatively undocumented web methods into the application, there's a self-describing endpoint that can be maintained/utilised when I'm gone)

Following the examples, I have successfully managed to build and embed CoreWCF into a test project that has it's roots in a standard .NET Core Console Application.

My program startup (for what it's worth) is:

```c#
            WebHost.CreateDefaultBuilder(args)
            .UseKestrel(options => { options.ListenLocalhost(8080); })
            .UseUrls("http://localhost:8080")
            .UseNetTcp(8808)
            .UseStartup<Startup>();
```

The middleware is doing its job, and I'm able to trace requests through to certain points in the CoreWCF project.  I'm not really following what's going on, and I'm not going to insult anyone's intelligence by pretending that I do.  I realise this is a new project, and that a lot of it probably isn't ready - but I want to do what I can to help move things forward by learning and understanding.

For this issue though, I've been trying to get Postman to return a message from an Operation Contract that has been made to work through a console application. 

I'm aware that out-of-the box, the SOAP version is tied to 1.1 (after many hours of trawling through various message samples and error messages, and finding the code in HTTPBinding!) 

The service contract I have created, is as follows:

```c#
    [ServiceContract]
    public interface ISOAPService
    {
        [OperationContract]
        string GetToken(string username, string passwordHash);
    }
```

Nothing strenuous at this point!

An example version 1.1 SOAP Message I have made to work 'the best' so far:

```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
 <soapenv:Header></soapenv:Header>
 <soapenv:Body>
 <GetToken>
	<Username>VorTechS</Username>
	<PasswordHash>abcdefghijkl</PasswordHash>
 </GetToken>
 </soapenv:Body>
</soapenv:Envelope>
```

I'm assuming the above is a correctly formatted message, with the appropriate request header set to text/xml to specify SOAP 1.1.

The problem is, that the message never gets mapped to associated operation contract.  Tracing down as far as I can through the code stack, it's because the message filtering is explicitly trying to find the Action from the Headers collection at this point:

`if (pairs[i].filter.Filter2.Match(message))` (line 258 of AndMessageFilterTable.cs)

Okay, so, perhaps I can just add an action?

```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
 <soapenv:Header>
 		<Action>http://localhost:8080/wcf/GetToken</Action>
</soapenv:Header>
 <soapenv:Body>
 <GetToken>
	<Username>VorTechS</Username>
	<PasswordHash>abcdefghijkl</PasswordHash>
 </GetToken>
 </soapenv:Body>
</soapenv:Envelope>
```

But this also fails to read an Action element.  Tracing through the code again, the code reading the headers, is return 255 at this point:

`kind = (HeaderKind)reader.IndexOfLocalName(localNames, version.Addressing.DictionaryNamespace);` (line 1317 of MessageHeaders.cs) for the Action element.

So for giggles I modified the Enum, and some associated code to deal with array sizes to get beyond this point only to be met with:

"The incoming message contains a SOAP header representing the WS-Addressing 'Action', yet the HTTP transport is configured with AddressingVersion.None.  As a result, the message is being dropped.  If this is not desired, then update your HTTP binding to support a different AddressingVersion."  which makes sense given the HTTP Binding is explicitly implementing this.

For even more giggles, I implemented a new shell MessageVersion for SOAP11 and Anonymous addressing.  

Which was fine, but then this meant the To header was missing, and that's because the code reading the header type for the To element is still returning 255, and so adding it as another Action header.

Is there a way I can construct a SOAP message that will correctly invoke my method out of the box?  Or should I keep digging?

**Comment by mconnew** [comment]
First, you don't need the UseNetTcp piece as POSTMAN (as far as I'm aware at least) only supports HTTP. I need to give you a quick primer on MessageVersion. The class MessageVersion encapsulates two concepts, the EnvelopeVersion and the AddressingVersion. The EnvelopeVersion is what specifies if you are using Soap 1.1 or Soap 1.2. The AddressingVersion specifies how the message address (which usually just means the Action although it can be more complicated) is specified. When using AdressingVersion.None, the Action gets put into HTTP headers. When using AddressingVersion.WSAddressing10 you are using the web standard WS-Addressing 1.0 which basically means put the address into the SOAP headers. The binding BasicHttpBinding by default uses MessageVersion.Soap11 which means EnvelopeVersion.Soap11 (So using Soap 1.1) and AddressingVersion.None which means the Action header is sent via HTTP headers. Anonymous addressing is something slightly different. It basically means the client doesn't provide a URL for the response to be sent to as the transport being used intrinsically has an unambiguous reply path for the request, i.e. HTTP requests have an HTTP response so just send the reply message back as the response that the transport (HTTP) provides.  

Looking at your example request SOAP message, you have the following at the start:
```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
```
The namespace there translates to you using Soap 1.1. You also have empty headers which tells me you are using addressing version none. You should have a SOAPAction header in your HTTP headers which is being used. The BasicHttpBinding should pull the action from the HTTP header and put it into the MessageHeaders of the incoming message so that the message filter code can work with it. It looks like you are debugging through the code to try work out what's going wrong. The place where this happens is in [HttpChannelHelpers.cs line 318](https://github.com/CoreWCF/CoreWCF/blob/master/src/CoreWCF.Http/src/CoreWCF/Channels/HttpChannelHelpers.cs#L318) (we read the HTTP header and store it in a local variable before fixing it up) and [line 347](https://github.com/CoreWCF/CoreWCF/blob/master/src/CoreWCF.Http/src/CoreWCF/Channels/HttpChannelHelpers.cs#L347) where we store the Action header into the incoming message headers. Hopefully this is enough info to get you unblocked.

**Comment by Kralizek** [comment]
Sorry for hijacking your issue but have you tried using the WCF test client (assuming you are working on windows). It's super convenient and does all the soap annoyance for you. You just have to make sure that your service publishes the metadata.

**Comment by VorTechS** [comment]
> 
> 
> Sorry for hijacking your issue but have you tried using the WCF test client (assuming you are working on windows). It's super convenient and does all the soap annoyance for you. You just have to make sure that your service publishes the metadata.

I appreciate the suggestion.  It's not publishing any metadata, hence why I was reverting to using Postman.  As I know the endpoint and had a vague idea of how to craft the request, just clearly not enough!

I think it's always good to have some sort of underlying knowledge of the requests when it comes to these things, even though I've shielded myself from it in the past - so this has been a useful exercise.

**Comment by mconnew** [comment]
@VorTechS, that sounds like it would be a good addition in the samples folder. We have an implementation of a server using CoreWCF, and the same service using .NET Framework as well as clients for .NET Core and .NET Framework. Adding a postman client would be great to add to the samples. It would need to run against the existing HTTP web service though so you would need to modify what you have to work with that contract.  
If you would like to contribute to the samples, feel free to open a PR.

**Comment by VorTechS** [comment]
Typically, having committed all that work, I decided to move on to trying out a complex type example.  

And whilst the appropriate method is being called, upon deserialization the method parameter properties are all null.  (The parameter value itself is an instantiated object of the appropriate type).

Example:

```
    [DataContract]
    public class BasicUserTokenRequest
    {
        public string Username { get; set; }

        public string PasswordHash { get; set; }
    }

[OperationContract]
 [ServiceKnownType(typeof(BasicUserTokenRequest))]
 string GetToken(BasicUserTokenRequest request);
```

...and the SOAP envelope modified to:

```
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
 <soapenv:Body>
 <GetToken xmlns="http://tempuri.org/">
 	<request>
		<Username>VorTechS</Username>
 		<PasswordHash>abcdefghijkl</PasswordHash>
  	</request>
  </GetToken>
  </soapenv:Body>
 </soapenv:Envelope>

```

I've traced things down to DataContractSerializationOperationFormatter.cs line 576, which then calls out to DataContractSerializer so I'm not able to see much further as to what the problem is.

Other queries around this subject in general suggest the envelope is correctly formatted, and certainly the deserialization part-way works.  

Am I missing something in the envelope that's stopping the additional properties from being read by the serializer?

**Comment by mconnew** [comment]
Create a WCF client using the same contract and see what the request looks like.  
But off the top of my head I can see a few issues.  I think `<GetToken>` might need to be `<GetTokenRequest>` and probably needs a namespace in there. `<request>` might been to be `<a:BasicUserTokenRequest xmlns:a="http://schemas.datacontract.org/2004/07/SomeNamespace>`. You don't generally need to create a request object to hold your parameters as WCF creates it for you with the operation name followed by Request. You can define your service like this:
```c#
  [OperationContract]
  string GetToken(string Username, string PasswordHash);
```
You only need ServiceKnownType if there's a type you need a serializer for which isn't immediately calculable from the method definition. For example, if you had a type derived from BasicUserTokenRequest which might get sent, you add that to the ServiceKnownType attribute to make sure the derived type can be deserialized too.

**Comment by VorTechS** [comment]
Thank you for the clarification of the usage of KnownType, I hadn't appreciated that before.

And apologies for the confusion around the variable 'request'.  It isn't an appropriately named variable in this instance!  It has nothing to do with the raw request.

As you suggested, I wrote a test using ChannelFactory:

```
            var myBinding = new BasicHttpBinding();
            var myEndpoint = new EndpointAddress(_soapEndPoint);

            using (var myChannelFactory = new ChannelFactory<ISOAPService>(myBinding, myEndpoint))
            {
                ISOAPService soapClient = null;

                soapClient = myChannelFactory.CreateChannel();
                soapClient.GetToken(new BasicUserTokenRequest() { Username = "VorTechS", PasswordHash = "Testing" });
                ((ICommunicationObject)soapClient).Close();
                myChannelFactory.Close();
            }
```

...and whilst the SOAP message isn't 1.1, the envelope at the point of Deserializing the Body Contents is apparently: 

```
<?xml version="1.0" encoding="utf-16"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Header>
    <To s:mustUnderstand="1" xmlns="http://schemas.microsoft.com/ws/2005/05/addressing/none">http://localhost:8080/wcf</To>
    <Action s:mustUnderstand="1" xmlns="http://schemas.microsoft.com/ws/2005/05/addressing/none">http://tempuri.org/ISOAPService/GetToken</Action>
  </s:Header>
  <s:Body>
    <GetToken xmlns="http://tempuri.org/">
      <request xmlns:a="http://schemas.datacontract.org/2004/07/UnitTests" xmlns:i="http://www.w3.org/2001/XMLSchema-instance">
      </request>
    </GetToken>
  </s:Body>
</s:Envelope>
```

Which appears to be missing the parameters!

So I looked further up the chain, right up as far as I could go to the RequestDelegateHandler, added some code to read the body contents from the HttpContent.Request ... and it's the same there!

```
<s:Envelope xmlns:s=\"http://schemas.xmlsoap.org/soap/envelope/\">
	<s:Body>
		<Register xmlns=\"http://tempuri.org/\"><request xmlns:a=\"http://schemas.datacontract.org/2004/07/PCGLionFishUnitTests\" xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\"/></Register>
	</s:Body>
</s:Envelope>
```

How curious!

**Comment by FreeAndNil** [comment]
Could the problem be that BasicUserTokenRequest is missing the DataMember-Attribute on its properties?

**Comment by mconnew** [comment]
Yes, that's exactly it. When you add the DataContract attribute, only properties with the DataMember attribute will be serialized. Basically the DataContract attribute turns all properties into explicit opt-in to be serialized. If you remove the DataContract attribute, it will serialize all public properties unless you explicitly opt-out by adding an attribute which I forget right now (I think NonSerializable maybe?). So either add DataMember or remove DataContract.

**Comment by VorTechS** [comment]
I'll double check that, but I'm sure I added the DataMember in my tests and that didn't change anything.  But I didn't try removing the DataContract property.

I did however

**Comment by VorTechS** [comment]
Sorry, I had a fat finger on mobile issue, didn't mean to close the issue!

Anyway, I did reproduce the issue on the sample projects - and will use those to continue getting to the bottom of things.  

It seems I can also get the server to stop responding to requests.  I'll update tomorrow when I've had chance to change things.

**Comment by VorTechS** [comment]
Hi all!  Okay.  So I got to the bottom of my issue, and there were TWO causes of the problem!

1. The Contract name in the xmlns of the 'request' node was wrong.  It needed to map to the root namespace of the application hosting the service.

**_2.  The parameters need to be in alphabetical order !?!?_**

Once the first issue had been addressed, the second showed only ONE of the parameters being populated!

For clarification, as you said above - if DataContract is used, DataMember must be also be used.  Both can be omitted though.

**Comment by FreeAndNil** [comment]
In classic WCF alphabetical order of properties is always required. I think the reason was performance while still being forward and backward compatible between different versions of a data contract.

**Comment by VorTechS** [comment]
> In classic WCF alphabetical order of properties is always required. I think the reason was performance while still being forward and backward compatible between different versions of a data contract.

I guess that makes some sense.  Does that also mean then that ALL properties are required in the envelope?

**Comment by FreeAndNil** [comment]
No, only properties with [DataMember(IsRequired = true)] are required by the serializer.

**Comment by mconnew** [comment]
There is an Order property on DataMember so you can explicitly specify the order. The xsd schema which correlates to DataContract classes expresses using a [sequence](https://www.w3schools.com/xml/el_sequence.asp) element which requires an order to be defined. The alternative would be to define the schema using an [all](https://www.w3schools.com/xml/el_all.asp) element. If you need to support any order, you need to use XmlSerializer instead, but there's a performance penalty as it's generally slower.

**Comment by VorTechS** [comment]
Thanks folks, it's all working nicely.  If it's of benefit, I've added an update to the examples showing an additional call using a complex object as a parameter.  I could update this to use the channel factory example too.  

I'll also double check, but expanding the examples in the way I did was causing the server to block.  I'll be honest, I'm not sure if that was just answering replies, or fully blocked, but I can check that out.
